Objects/typeobject.c

```diff
@@ -3384,6 +3390,12 @@ PyType_FromModuleAndSpec(PyObject *module, PyType_Spec *spec, PyObject *bases)
     char *res_start;
     short slot_offset, subslot_offset;
 
+    if (meta->tp_new != PyType_Type.tp_new) {
```

Comment:
n.b.: If numpy were to use this to migrate to heap types for the dtypes, their `DTypeMeta` class currently does define a custom `tp_new`:
```
dtypemeta_new(PyTypeObject *NPY_UNUSED(type),
        PyObject *NPY_UNUSED(args), PyObject *NPY_UNUSED(kwds))
{
    PyErr_SetString(PyExc_TypeError,
            "Preliminary-API: Cannot subclass DType.");
    return NULL;
}
```

I'm not sure how important it would be for them to keep this error message, maybe someone from NumPy can comment?

See: https://github.com/numpy/numpy/blob/ae8b9ce90dc2d5922def5db8cf7d4410c60b3e13/numpy/core/src/multiarray/dtypemeta.c#L45

Reply:
Disallowing subclassing can already be achieved by clearing the `Py_TPFLAGS_BASETYPE` flag:

https://docs.python.org/3/c-api/typeobj.html#Py_TPFLAGS_BASETYPE

https://github.com/python/cpython/blob/079f0dd7191fbadd4c3a5899b6af12492e84d2b4/Objects/typeobject.c#L3502-L3508

Reply:
Yeah, disallowing subclassing might work (of course the dtypes are subclasses, but I guess we don't disallow subclasses in C with that).
In practice, I don't expect NumPy needs a `tp_new`.  I would love to see a dataclass-like API to create a new DType at some point, but that would work differently anyway.  So this should be OK.

I would love if @encukou would comment on this, because I think he has a clearer picture than me.

From my current understanding, there were two scenarios:
1. This checks makes sense for the _current_ API where a metaclass may exist in the bases!  (a current bug that this is ignored, though!).  Since we may otherwise forget to initialize the metaclass properly.
2. For the *new* function, it might make sense to tell users that it is *their* responsibility to make sure the metaclass is properly initialized, and just not add this check at all!

Now, Petr had suggested `ApplySpec` for the second use-case.  I am not sure if that is as convenient in practice (why not call alloc for the user?).  But the point is, the new function is for metaclass providers who know what additional initialization is needed for their metaclass (so the check isn't really needed).